src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,141 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] internalArray;
+    private int size;
+
+    public ArrayList() {
+        this.internalArray = new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        this.internalArray = new Object[capacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size != internalArray.length) {
+            for (int i = 0; i < internalArray.length; i++) {
+                if (i == size) {
+                    internalArray[i] = value;
+                    size++;
+                    break;
+                }
+            }
+        } else {
+            grow();
+            internalArray[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            growByCell(index);
+            internalArray[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index!!!");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            return (T) this.internalArray[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("The such index doesn't exist");
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < size && index >= 0) {
+            internalArray[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong array index!");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T oldelement = null;
+        if (index < this.size && index >= 0) {
+            int subsiquentIndex = 0;
+            oldelement = get(index);
+            Object[] newFormedArray = new Object[internalArray.length];
+            for (int i = 0; i < size; i++) {
+                if (i != index) {
+                    newFormedArray[subsiquentIndex] = internalArray[i];
+                    subsiquentIndex++;
+                }
+            }
+            internalArray = newFormedArray;
+            size--;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index");
+        }
+        return oldelement;
```

Comment:
Use System.arraycopy() to move your array elements instead of manually copying them in a loop. Also, avoid creating a new array every time you remove an element. Instead, you can shift the remaining elements to the left to fill the gap.